SwiftCompilerSources/Sources/AST/DiagnosticEngine.swift

```diff
@@ -0,0 +1,117 @@
+//===--- DiagnosticEngine.swift -------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+@_exported import ASTBridging
+import Basic
+
+public typealias DiagID = BridgedDiagID
+
+extension BridgedDiagnosticArgument {
+  init(stringRef val: BridgedStringRef) {
+    self.init(kind: .stringRef, value: .init(stringRefValue: val))
+  }
+  init(int val: Int) {
+    self.init(kind: .int, value: .init(intValue: val))
+  }
+  init(uInt val: UInt) {
+    self.init(kind: .uInt, value: .init(uintValue: val))
+  }
+}
+
+public protocol DiagnosticArgument {
+  func withBridgedDiagnosticArgument(_ fn: (BridgedDiagnosticArgument) -> Void)
+}
+extension String: DiagnosticArgument {
+  public func withBridgedDiagnosticArgument(_ fn: (BridgedDiagnosticArgument) -> Void) {
+    withBridgedStringRef { fn(BridgedDiagnosticArgument(stringRef: $0)) }
+  }
+}
+extension Int: DiagnosticArgument {
+  public func withBridgedDiagnosticArgument(_ fn: (BridgedDiagnosticArgument) -> Void) {
+    fn(BridgedDiagnosticArgument(int: self))
+  }
+}
+extension UInt: DiagnosticArgument {
+  public func withBridgedDiagnosticArgument(_ fn: (BridgedDiagnosticArgument) -> Void) {
+    fn(BridgedDiagnosticArgument(uInt: self))
+  }
+}
+
+public struct DiagnosticFixIt {
+  public var start: SourceLoc
+  public var byteLength: Int
+  public var text: String
+
+  public init(start: SourceLoc, byteLength: Int, replacement text: String) {
+    self.start = start
+    self.byteLength = byteLength
+    self.text = text
+  }
+
+  func withBridgedDiagnosticFixIt(_ fn: (BridgedDiagnosticFixIt) -> Void) {
+    text.withBridgedStringRef { bridgedTextRef in
+      let bridgedDiagnosticFixIt = BridgedDiagnosticFixIt(
+        start: start.bridged,
+        byteLength: byteLength,
+        text: bridgedTextRef)
+      fn(bridgedDiagnosticFixIt)
+    }
+  }
+}
+
+public struct DiagnosticEngine {
+  private var bridged: BridgedDiagnosticEngine
+
+  public init(bridged: BridgedDiagnosticEngine) {
+    self.bridged = bridged
+  }
+
+  public func diagnose(_ position: SourceLoc?,
+                       _ id: DiagID,
+                       _ args: [DiagnosticArgument],
```

Comment:
Can you make this a var-arg parameter instead of an array?

Reply:
I added a var-arg variant, but I still want `Array` one too. It's convenient when, for example, using Swift's enum as an diagnostic type.
For example: 
```swift
struct ParserDiagnostic {
  enum Message {
    case expectedNumDigits(String, Int)
    case miscError(String)
    case otherError
  }
  let location: Location
  let message: Message
}

func reportDiag(diag: ParseError) {
  let location = getSourceLoc(location)
  let diagID: DiagID
  let arguments: [DiagnosticArguement]

  switch diag.message {
  case expectedNumDigits(let s, let i):
    diagID = .parse_regex_num_digits
    arguments = [s, I]

  case ...
    ...
  }
  
  engine.diagnose(loc, diagID, arguments)
}
```
Otherwise, we need to call `DiagnositcEngine.diagnose()` from each `case`